Gorilla column encoding

Enmk opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):
DoubleDelta and Gorilla  column codecs
...

Detailed description (optional):
- DoubleDelta column codec implementation is based on Gorilla paper: http://www.vldb.org/pvldb/vol8/p1816-teller.pdf, which was extended to support 64bit types. The drawback is 1 extra bit for 32-byte wide deltas: 5-bit prefix instead of 4-bit prefix.

- Gorilla column codec implementation is based on Gorilla paper: http://www.vldb.org/pvldb/vol8/p1816-teller.pdf, which was extended to support 64bit types. The base use-case for the codec is floating-point values that do not change rapidly.
...

Enmk committed:
DoubleDelta column encoding.

Added DoubleDelta column encoding + test cases;
Added BitWriter and BitReader that allow to read data from ReadBuffer
and write data to WriteBuffer bit by bit, up to 64 bits at once;
Added test for BitReader and BitWriter.

Enmk committed:
Gorilla column encoding

Added Gorilla column encoding for any fixed-width type;
Added tests for Float32 and Float64 values.

Enmk committed:
Docs for DoubleDelta and Gorilla column codecs.

4ertus2 commented on dbms/src/Compression/CompressionCodecDoubleDelta.cpp:
```diff
@@ -0,0 +1,320 @@
+        else
+        {
+            const auto sign = std::signbit(double_delta);
+            const auto abs_value = static_cast<typename std::make_unsigned<DeltaType>::type>(std::abs(double_delta));
```
```
using UnsignedDeltaType = typename std::make_unsigned<DeltaType>::type;
```
in front of function

4ertus2 replied:
Also what about INT_MIN? abs(INT_MIN) == INT_MIN.

Enmk replied:
Please take a look at the unit tests I've added. `dbms/src/Compression/tests/gtest_compressionCodec.cpp`

4ertus2 commented on dbms/src/Compression/CompressionCodecDoubleDelta.cpp:
```diff
@@ -0,0 +1,320 @@
+        {
+            const auto sign = std::signbit(double_delta);
+            const auto abs_value = static_cast<typename std::make_unsigned<DeltaType>::type>(std::abs(double_delta));
+            if (double_delta > -63 && double_delta < 64)
```
it's better to make a function
(-64, 64) -> 0b10
(-255, 255) -> 0b110
...
It also should return needed_bits as result. Then we would have no copy-paste blocks here.

Enmk replied:
Done

4ertus2 commented on dbms/src/Compression/CompressionCodecDoubleDelta.cpp:
```diff
@@ -0,0 +1,320 @@
+    T prev_value{};
+    DeltaType prev_delta{};
+
+    if (source < source_end)
```
The same it's better to check before the call.

4ertus2 commented on dbms/src/Compression/CompressionCodecDoubleDelta.cpp:
```diff
@@ -0,0 +1,320 @@
+        DeltaType double_delta = 0;
+        if (reader.readBit() == 0)
+        {
+            double_delta = 0;
```
It looks strange. What's the difference with
```
if (reader.readBit() != 0)
{
    /// all the things in lese block?
}
```

Enmk replied:
That's a good idea.

Enmk replied:
Done

4ertus2 commented on dbms/src/Compression/CompressionCodecDoubleDelta.cpp:
```diff
@@ -0,0 +1,320 @@
+        else
+        {
+            // first bit is 1
+            const UInt8 data_sizes[] = {6, 8, 11, 31, 63};
```
Move outside of cycle

Enmk replied:
I really doubt that will add any performance or readability.

4ertus2 commented on dbms/src/Compression/CompressionCodecDoubleDelta.cpp:
```diff
@@ -0,0 +1,320 @@
+            // first bit is 1
+            const UInt8 data_sizes[] = {6, 8, 11, 31, 63};
+            UInt8 i = 0;
+            for (; i < sizeof(data_sizes) - 1; ++i)
```
It looks like a slow solution.

Enmk replied:
'Looks' is a funny way to judge about performance ;) 
Anyway, what would you suggest me to do? Unroll a loop, construct a lookup-table? Anything else?

4ertus2 replied:
I think we could implement countLeadingZeroes() and countLeadingOnes() for BitReader and try to optimize them.

Enmk replied:
Am I reading this correct: no modifications to this part are required right now ?

4ertus2 replied:
I'd prefer you extract this cycle into BitReader function for now and give it a good name.

Enmk replied:
Moving code around wouldn't make it any faster.

Enmk replied:
Well, I ran extra mile to optimize `BitReader` and `BitWriter` - now they have `UInt64`-buffers instead of `UInt8`, that means 8 times less calls to underlying buffer for reading or writing more data. IMO this should be performant enough now.

4ertus2 commented on dbms/src/Compression/CompressionCodecDoubleDelta.cpp:
```diff
@@ -0,0 +1,320 @@
+        compressed_size = compressDataForType<UInt32, Int64>(&source[bytes_to_skip], source_size - bytes_to_skip, &dest[start_pos]);
+        break;
+    case 8:
+        compressed_size = compressDataForType<UInt64, Int64>(&source[bytes_to_skip], source_size - bytes_to_skip, &dest[start_pos]);
```
Is it enough to have Int64 for any combinations of UInt64? Do we throw out of range exception otherwise?

Enmk replied:
I think that I have a test case just for that, could you please validate it? https://github.com/yandex/ClickHouse/pull/5600/files#r293377862

4ertus2 commented on dbms/src/Compression/CompressionCodecDoubleDelta.cpp:
```diff
@@ -0,0 +1,320 @@
+        decompressDataForType<UInt32, Int64>(&source[2 + bytes_to_skip], source_size_no_header, &dest[bytes_to_skip]);
+        break;
+    case 8:
+        decompressDataForType<UInt64, Int64>(&source[2 + bytes_to_skip], source_size_no_header, &dest[bytes_to_skip]);
```
The same question.

Enmk commented on dbms/src/Compression/CompressionCodecDoubleDelta.cpp:
```diff
@@ -0,0 +1,320 @@
+    T prev_value{};
+    DeltaType prev_delta{};
+
+    if (source < source_end)
```
Well, it is not about zero-sized source, we're constantly moving the pointer, and later verify that we're not reading from outside of buffer.

Does it make any sense to wrap that raw memory into a `ReadBuffer` ? And avoid manual boundary checks altogether?

4ertus2 replied:
OK, understand. It doesn't make sense. I think for codecs we should use more C-style with clear memory management.

Enmk commented on dbms/src/Common/BitHelpers.h:
```diff
@@ -30,3 +31,32 @@ inline size_t roundUpToPowerOfTwoOrZero(size_t n)
+
+
+template <typename T>
+inline std::enable_if_t<std::is_integral_v<T> && (sizeof(T) <= sizeof(unsigned int)), int>
```
Agree

Enmk commented on dbms/tests/queries/0_stateless/00950_column_encoding_double_delta.sql:
```diff
@@ -0,0 +1,211 @@
+INSERT INTO reference (key, valueU64, valueU32, valueU16, valueU8, valueI64, valueI32, valueI16, valueI8, valueDT, valueD)
+	SELECT number as n, n as v, v, v, v, v, v, v, v, toDateTime(v), toDate(v) FROM system.numbers LIMIT 101, 100;
+
+-- checking for overflow
```
Overflow test cases

4ertus2 replied:
18446744073709551616 is String to UInt64 overflow. I expected to see 0xffffffffffffffff, 0x7fffffffffffffff,0x8000000000000000 in cases where overflow is inside compression func not in input phase. Ex. 
```
0x7fffffffffffffff - (-1)
0x7fffffffffffffff - 0x8000000000000000
```

Enmk replied:
This issue was addresses in unit tests.

Enmk committed:
Post-PR fixes:

* BitHelpers.cpp was removed, corresponding code was moved to the header
* BitIO test as GTest-based test binary
* gtest-based unit test for DoubleDelta and Gorilla codecs
* getLeadingZeroBits from SFINAE to consexpr if
* removed couple of unneeded if's
* Fixed sql-test to use one table

4ertus2 commented on dbms/tests/queries/0_stateless/00950_column_encoding_double_delta.sql:
```diff
@@ -0,0 +1,211 @@
+	r.valueU64, d.valueU64, r.valueU64 - d.valueU64 as dU64
+FROM reference as r, doubleDelta as d
+WHERE
+	r.key == d.key
```
It does not test if data in tables is the same but some values are mixed. It's better to avoid JOIN here and insert both raw and DeltaDelta values into one table and get results as
```
select * from t where i64 != i64_delta or i32 != i32_delta ...
```

Enmk replied:
Done

Enmk commented on dbms/src/Compression/CompressionCodecDoubleDelta.cpp:
```diff
@@ -0,0 +1,320 @@
+    while (source < source_end)
+    {
+        const T curr_value = unalignedLoad<T>(source);
+        source += sizeof(curr_value);
```
This kinda breaks the pattern that is used everywhere else:
```
const T curr_value = unalignedLoad<T>(source);
source += sizeof(curr_value);
```

alexey-milovidov commented:
https://clickhouse-test-reports.s3.yandex.net/5600/d1904c251ee29547eb12bb2e5b7a85379b992739/pvs_studio_report/pvs-studio-html-report/sources/BitHelpers.cpp_32.html#ln102

Enmk committed:
Post-PR fixes #2

Fixed style issues and build for clang-7;
64-bit buffer for BitReader and BitWriter;
Fixed overflow and writing more bytes on flushing;
Added maskLowBits() and tests for it.

Enmk commented on dbms/src/Compression/tests/gtest_compressionCodec.cpp:
```diff
@@ -0,0 +1,315 @@
+    )
+);
+
+INSTANTIATE_TEST_CASE_P(Overflow,
```
Overflow test cases for `DoubleDelta` and `Gorilla` codecs.

Enmk committed:
Post-PR fixes stage 3

maskLowBits()
building for clang
more test cases for codecs.

Enmk committed:
Updated test cases for gorilla to simplify debugging

Added case name and previous row values.

Enmk committed:
Fixed UB in maskLowBits

Enmk committed:
Performance tests for Gorilla and DoubleDelta

Testing INSERTs and SELECTs on codec combinations:
 * DoubleDelta (or Gorilla)
 * LZ4
 * DD (or G) + LZ4
On 3 types of data:
 * sequential
 * monotonic
 * random.

alesapin commented on dbms/tests/performance/codec_gorilla.xml:
```diff
@@ -0,0 +1,71 @@
+    <fill_query>INSERT INTO rnd_{table_suffix} (n) SELECT n FROM rnd_{table_suffix}</fill_query>
+
+
+    <query>INSERT INTO seq_{table_suffix} (n) SELECT number/pi() FROM system.numbers SETTINGS max_threads=1</query>
```
This query will be run in a loop and will be compared on the `min_time` metric. This query is endless, it doesn't have `min_time`. You can see MAX_INT numbers in comparison report https://clickhouse-test-reports.s3.yandex.net/5600/7d38038964faf624c04ab1eb9bcaa6c47b6a02be/performance_test/comparison_min_time_gcc_8.html . You have to `limit` numbers table to some number or change type of test to `once`. See https://github.com/yandex/ClickHouse/blob/master/dbms/tests/performance/README.md.
